libbeat/processors/processor.go

```diff
@@ -36,35 +41,53 @@ type Processor interface {
 }
 
 func New(config PluginConfig) (*Processors, error) {
-	procs := Processors{}
-
-	for _, processor := range config {
+	procs := &Processors{
+		log: logp.NewLogger(logName),
+	}
 
-		if len(processor) != 1 {
-			return nil, fmt.Errorf("each processor needs to have exactly one action, but found %d actions",
-				len(processor))
+	for _, procConfig := range config {
+		// Handle if/then/else processor which has multiple top-level keys.
+		if procConfig.HasField("if") {
+			p, err := NewIfElseThenProcessor(procConfig)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to make if/then/else processor")
+			}
+			procs.add(p)
+			continue
 		}
 
-		for processorName, cfg := range processor {
+		if len(procConfig.GetFields()) != 1 {
+			return nil, errors.Errorf("each processor must have exactly one "+
+				"action, but found %d actions (%v)",
+				len(procConfig.GetFields()),
+				strings.Join(procConfig.GetFields(), ","))
+		}
 
-			gen, exists := registry.reg[processorName]
-			if !exists {
-				return nil, fmt.Errorf("the processor %s doesn't exist", processorName)
-			}
+		actionName := procConfig.GetFields()[0]
+		actionCfg, err := procConfig.Child(actionName, -1)
+		if err != nil {
+			return nil, err
+		}
 
-			cfg.PrintDebugf("Configure processor '%v' with:", processorName)
-			constructor := gen.Plugin()
-			plugin, err := constructor(cfg)
-			if err != nil {
-				return nil, err
-			}
+		gen, exists := registry.reg[actionName]
+		if !exists {
+			return nil, errors.Errorf("the processor action %s does not exist", actionName)
+		}
 
-			procs.add(plugin)
+		actionCfg.PrintDebugf("Configure processor action '%v' with:", actionName)
+		constructor := gen.Plugin()
+		plugin, err := constructor(actionCfg)
+		if err != nil {
+			return nil, err
 		}
+
+		procs.add(plugin)
 	}
 
-	logp.Debug("processors", "Processors: %v", procs)
-	return &procs, nil
+	if len(procs.List) > 0 {
+		procs.log.Debugf("Generated new processors: %v", procs)
+	}
+	return procs, nil
```

Comment:
This will only count the top level processors and not the processors defined inside a If/Else/branch?

Reply:
It should log information about all of the processors, even nested ones. But since there are multiple calls to `New` you will see a few of these debug statements when using if/then/else. The last one being the complete processor tree.

```
=== RUN   TestIfElseThenProcessor/if-then-else-if
2019-02-14T10:04:42.904-0500	DEBUG	[conditions]	conditions/conditions.go:95	New condition range: map[uid:{<nil> <nil> <nil> 0xc0000a6dc0}]
2019-02-14T10:04:42.904-0500	DEBUG	[processors]	processors/processor.go:88	Generated new processors: add_fields={"uid_type":"reserved"}
2019-02-14T10:04:42.904-0500	DEBUG	[conditions]	conditions/conditions.go:95	New condition equals: map[uid:{500  false}]
2019-02-14T10:04:42.904-0500	DEBUG	[processors]	processors/processor.go:88	Generated new processors: add_fields={"uid_type":"eq_500"}
2019-02-14T10:04:42.904-0500	DEBUG	[processors]	processors/processor.go:88	Generated new processors: add_fields={"uid_type":"gt_500"}
2019-02-14T10:04:42.904-0500	DEBUG	[processors]	processors/processor.go:88	Generated new processors: if equals: map[uid:{500  false}] then add_fields={"uid_type":"eq_500"} else add_fields={"uid_type":"gt_500"}
2019-02-14T10:04:42.904-0500	DEBUG	[processors]	processors/processor.go:88	Generated new processors: if range: map[uid:{<nil> <nil> <nil> 0xc0000a6dc0}] then add_fields={"uid_type":"reserved"} else if equals: map[uid:{500  false}] then add_fields={"uid_type":"eq_500"} else add_fields={"uid_type":"gt_500"}
```